### ISSUE TITLE
lin

### ISSUE THREAD

**Issue by babypure** [Lyn-Wu/Linker_Project#1]
，

**Comment by Lyn-Wu** [comment]
您好，这是我为了熟悉链接器写的 水平有限

**Comment by babypure** [comment]
您好，感谢回复。
我看了你之前复现的D2NN的code，并run了一下，结果似乎不收敛，您po出的两张图的得分看图也并非对应每个字符的识别率，可以麻烦您解释一下吗？包括propagation的return是否应该是E0乘本征函数乘相位吗？看您的code似乎又不像？
期待收到您的回复！
也可以加wechat聊: xcw1123

Best wishes,
Xuechun

Sent from my iPhone

On 31 Jul 2019, at 03:45, Lyn <notifications@github.com<mailto:notifications@github.com>> wrote:


您好，这是我为了熟悉链接器写的 水平有限

**Comment by Lyn-Wu** [comment]
首先 里面的结果是修改final.py的波长训练出来的时候原论文是用太赫兹波，而我们尝试用红外激光，其次propagation用的是角谱法

**Comment by Lyn-Wu** [comment]
使用论文上的参数效果很好，当使用氦氖激光（跟正 刚才写错了）需要比较难实现的条件才可以，
为了实验上能做出来，所以修改了一些参数

**Comment by Lyn-Wu** [comment]
![Uploading IMG_0431.JPG…]()
![IMG_0432](https://user-images.githubusercontent.com/35756813/62181738-d6d56200-b343-11e9-9be3-b5843047d798.JPG)
![IMG_0433](https://user-images.githubusercontent.com/35756813/62181741-d76df880-b343-11e9-823c-aac8768927ff.JPG)

**Comment by Lyn-Wu** [comment]
您可以看一下当时按论文参数 得出来的结果

**Comment by Lyn-Wu** [comment]
![IMG_0431](https://user-images.githubusercontent.com/35756813/62181856-2f0c6400-b344-11e9-8a97-4399b5e0aea6.JPG)
![IMG_0432](https://user-images.githubusercontent.com/35756813/62181857-329feb00-b344-11e9-9bc2-e439cf60c7c0.JPG)
![IMG_0433](https://user-images.githubusercontent.com/35756813/62181867-3c295300-b344-11e9-8e01-d383e1edf2df.JPG)

**Comment by babypure** [comment]
我看你那个code的phase是定值，您有phase-optimize的code吗？
还有您定义了theta但是没有用，是有什么用处呢？
我跑了您的code, 但是识别率曲线是这样，是因为每次都random取初值所以结果如此吗？不应该趋于定值吗？
[cid:7b18b308-df03-4787-b19a-939a20b5ce8f]
________________________________

**Comment by babypure** [comment]
我也有看过另一个paper, propagation也是用的角谱法(第一页)，但不相同？
https://arxiv.org/abs/1811.03370

期待收到您的回复 蟹蟹~


________________________________

**Comment by Lyn-Wu** [comment]
实验开始之前，我们设计调幅和调相位两种，后来只采取了一种就没写。所以为常量。
关于角谱法，我们还做了Fourier变换

**Comment by babypure** [comment]
你们有木有硬件实现？前面提到使用论文参数效果很好 只是改动了Hz是吗？
如果想做phase调制，是不是再加一个phase和距离z对应的关系就可以？我现在困惑于如何做phase optimize, 可否提供一条思路？
________________________________

**Comment by babypure** [comment]
同时，如何解释这个loss和精度结果？(没有对code做改动)
________________________________

**Comment by babypure** [comment]
您好~期待收到您的回复
________________________________

**Comment by Lyn-Wu** [comment]
您看下All-optical-machine-learning-using-diffractive-deep-neural-networks 论文，loss函数就是依照上面介绍的写的

**Comment by babypure** [comment]
emm..你的code是振幅优化，用的也是paper的loss（振幅平方），我觉得理解起来没问题。
但是loss和phase没有直接关系？如果做optimize-phase的话，phase是怎么更新的呢？
[cid:23b84c43-2768-49bc-9adf-3c773af12ec8]
________________________________

**Comment by babypure** [comment]
另：想搞清你的角谱法code的具体含义。
tf.ifft2d(np.fft.fftshift(H)*tf.fft2d(u0)*dL*dL/(N*N))*N*N*df*df
我看了一些角谱法的理论 但是感觉和code对应不起来。

http://xbbjb.swu.edu.cn/viscms/u/cms/xbbjb/201410/21080853ju4g/2014-09-118.pdf

________________________________

**Comment by Lyn-Wu** [comment]
[????????????????_???.pdf](https://github.com/Lyn-Wu/Linker_Project/files/3467964/_.pdf)
角谱理论用于衍射光学元件数字仿真_殷可为

**Comment by babypure** [comment]
谢谢，完全搞明白了。
但是对于结果还是不收敛，改了learning rate为0.001也不收敛，请问是什么原因？c = 3e8
Hz = 0.4e12也是用的paper的参数。打算再改batch size参数为32试试？
________________________________

**Comment by Lyn-Wu** [comment]
32可能不够 我记得当时我设置64和128

**Comment by babypure** [comment]
我跑了64 但是不收敛 你有碰到不收敛问题吗
________________________________

**Comment by Lyn-Wu** [comment]
用论文的参数 没有

**Comment by babypure** [comment]
论文参数是每层200*200神经元 每层大小8cm epoch=8
你当时是用这个参数做的吗
________________________________

**Comment by Lyn-Wu** [comment]
神经元的数量不一样

**Comment by babypure** [comment]
learning_rate = 0.001
size = 512​
delta = 0.03 ​
dL = 0.02 ​
batch_size = 64​
batch = 10 ​
c = 3e8​
Hz = 0.4e12
你是用这个参数做的没问题是吗 非常希望得到您的解答 因为我换了参数依然不收敛 不明白问题出在哪里 code没有改动
________________________________

**Comment by Lyn-Wu** [comment]
把 delta和dL改为论文论文的参数 试一试

**Comment by Lyn-Wu** [comment]
波长不要变

**Comment by Lyn-Wu** [comment]
我记得当时2epoch就能到85～%

**Comment by babypure** [comment]
好 感谢！我再试试~
________________________________

**Comment by babypure** [comment]
我用了paper的每层大小0.08，间距0.03，但是结果还是不对。
想确定下code完全没问题~感谢！
________________________________

**Comment by Lyn-Wu** [comment]
什么结果不对？
![1](https://user-images.githubusercontent.com/35756813/62629713-4a88f780-b960-11e9-80d6-58d744be28ae.PNG)
![2](https://user-images.githubusercontent.com/35756813/62629715-4c52bb00-b960-11e9-9e30-15a68923f313.PNG)
![3](https://user-images.githubusercontent.com/35756813/62629720-4e1c7e80-b960-11e9-9343-6c392776153b.PNG)
![4](https://user-images.githubusercontent.com/35756813/62629725-4fe64200-b960-11e9-8a5c-c5d591b5c447.PNG)

**Comment by Lyn-Wu** [comment]
请把不对的结果详细介绍一下，不要说一个不对。你可以写一个测试程序看利用训练的结果是否真确

**Comment by babypure** [comment]
你的训练结果没问题。loss一直减小，但是我训练20次，loss不是一直减小的，一直不明白这个原因。

epoch :1 loss:7.8090 acc:0.7812
epoch :2 loss:5.4375 acc:0.8750
epoch :3 loss:8.2036 acc:0.7812
epoch :4 loss:2.1262 acc:0.9375
epoch :5 loss:1.0475 acc:0.9688
epoch :6 loss:5.5368 acc:0.8438
epoch :7 loss:3.2046 acc:0.9062
epoch :8 loss:3.4682 acc:0.9062
epoch :9 loss:8.0621 acc:0.8125
epoch :10 loss:1.0601 acc:0.9688
epoch :11 loss:2.3777 acc:0.9375
epoch :12 loss:1.1268 acc:0.9688
epoch :13 loss:1.1272 acc:0.9688
epoch :14 loss:3.5093 acc:0.9375
epoch :15 loss:2.1316 acc:0.9375
epoch :16 loss:2.0309 acc:0.9688
epoch :17 loss:7.5594 acc:0.8438
epoch :18 loss:2.7808 acc:0.9375
epoch :19 loss:2.1393 acc:0.9375
epoch :20 loss:1.1043 acc:0.9688
________________________________

**Comment by Lyn-Wu** [comment]
learning_rate=0.01 试试

**Comment by babypure** [comment]
上面的结果设置的参数是：
learning_rate = 0.01
size = 512
delta = 0.03
dL = 0.08
batch_size = 32
batch = 10
c = 3e8；Hz = 0.4e12
paper里的batch_size是8，但是鉴于你用64的也是收敛，我认为问题不在这个。
________________________________

**Comment by babypure** [comment]
认为不收敛要么是模型问题，要么是设置参数问题。
模型私认为没问题，但参数的一点变化就会导致不收敛吗..
另：针对phase optimize, 是不是output = propagation*Amp*exp(jφ)？把φ也设置成随机值输入。
________________________________

**Comment by babypure** [comment]
或者您可以再试一下用那个code是否收敛?
改了很多次参数了lr = 0.001也是不收敛，
________________________________

**Comment by babypure** [comment]
另：关于loss
loss = tf.reduce_sum(tf.square(logits_abs data_y),reduction_indices=[1])
这里为什么会用reduce_sum，reduce_sum不是降维累加吗？
________________________________

**Comment by babypure** [comment]
我改成了loss = tf.reduce_mean(tf.square(logits_abs data_y)，batch_size改成了paper里的8，结果更加不收敛
________________________________

**Comment by babypure** [comment]
期待收到您的回复，您用的全部参数是多少。
________________________________

**Comment by babypure** [comment]
我又试了每层大小0.08/0.03, batch_size128, 均不收敛。期待收到您的回复。
epoch :1 loss:0.1043acc:0.9375
epoch :2 loss:0.0227acc:0.9922
epoch :3 loss:0.0555 acc:0.9688
learning_rate = 0.01
delta = 0.03
dL = 0.02
batch_size = 128

epoch :1 loss:0.1642 acc:0.9062
epoch :2 loss:0.0244 acc:0.9688
epoch :3 loss:0.1250 acc:0.9375
learning_rate = 0.01
delta = 0.03
dL = 0.08
batch_size = 32

epoch :1 loss:0.0004 acc:1.0000
epoch :2 loss:0.1118 acc:0.9375
epoch :3 loss:0.1874 acc:0.9062
epoch :4 loss:0.1875 acc:0.9062
learning_rate = 0.01
delta = 0.03
dL = 0.02
batch_size = 32


________________________________

**Comment by babypure** [comment]
您方便查看一下这个code和之前跑的可行的code有哪些区别吗~蟹蟹
________________________________

**Comment by babypure** [comment]
期待收到您的回复
________________________________

**Comment by babypure** [comment]
另：是不是应该用远场算法，我改成了远场算法，但是一直报错，not enough values to unpack (expected 2, got 0)，你之前碰到过这个问题吗？
________________________________
